src/java/org/apache/cassandra/db/marshal/NativeAccessor.java

```diff
@@ -0,0 +1,447 @@
+    {
+        dstAccessor.copyByteBufferTo(src.asByteBuffer(), srcOffset, dst, dstOffset, size);
+        return size;
+    }
+
+    @Override
+    public int copyByteArrayTo(byte[] src, int srcOffset, NativeData dstNative, int dstOffset, int size)
+    {
+        ByteBuffer dst = dstNative.asByteBuffer();
+        FastByteOperations.copy(src, srcOffset, dst, dst.position() + dstOffset, size);
+        return size;
+    }
+
+    @Override
+    public int copyByteBufferTo(ByteBuffer src, int srcOffset, NativeData dstNative, int dstOffset, int size)
+    {
+        ByteBuffer dst = dstNative.asByteBuffer();
+        FastByteOperations.copy(src, src.position() + srcOffset, dst, dst.position() + dstOffset, size);
+        return size;
+    }
+
+    @Override
+    public void digest(NativeData value, int offset, int size, Digest digest)
+    {
+        ByteBuffer byteBuffer = value.asByteBuffer();
+        digest.update(byteBuffer, byteBuffer.position() + offset, size);
+    }
+
+    @Override
+    public NativeData slice(NativeData input, int offset, int length)
+    {
+        return input.slice(offset, length);
+    }
+
+    @Override
+    public <VR> int compare(NativeData left, VR right, ValueAccessor<VR> accessorR)
+    {
+        if (right instanceof NativeData)
+        {
+            return left.compareTo((NativeData) right);
+        }
+        return left.compareTo(accessorR.toBuffer(right));
+    }
+
+    @Override
+    public int compareByteArrayTo(byte[] left, NativeData right)
+    {
+        return ByteBufferUtil.compare(left, right.asByteBuffer());
+    }
+
+    @Override
+    public int compareByteBufferTo(ByteBuffer left, NativeData right)
+    {
+        return -right.compareTo(left); // we want to avoid ByteBuffer retrieval from NativeData
+    }
+
+     // -----------------------------------------------------------------------------
+     // Data deserialization methods
+
+    @Override
+    public byte[] toArray(NativeData value)
+    {
+        if (value == null)
+            return null;
+        return ByteBufferUtil.getArray(value.asByteBuffer());
+    }
+
+    @Override
+    public byte[] toArray(NativeData value, int offset, int length)
+    {
+        if (value == null)
+            return null;
+        ByteBuffer byteBuffer = value.asByteBuffer();
+        return ByteBufferUtil.getArray(byteBuffer, byteBuffer.position() + offset, length);
+    }
+
+    @Override
+    public String toString(NativeData value, Charset charset) throws CharacterCodingException
+    {
+        return ByteBufferUtil.string(value.asByteBuffer(), charset);
+    }
+
+    @Override
+    public String toHex(NativeData value)
+    {
+        return ByteBufferUtil.bytesToHex(value.asByteBuffer());
+    }
+
+    @Override
+    public byte toByte(NativeData value)
+    {
+        return ByteBufferUtil.toByte(value.asByteBuffer());
+    }
+
+    @Override
+    public byte getByte(NativeData value, int offset)
```

Comment:
`NativeData` should have an address and length, and then all these can be implemented using `MemoryUtil`'s methods.

Reply:
agree, I am working on it.
an implementation note: the current MemoryUtil methods uses native order, while CQL specification (as well as buffer wrappers exposed in the current trunk logic) uses BIG_ENDIAN, so I am adding explicit BIG_ENDIAN methods to MemoryUtil

Reply:
working on a test coverage